src/lib/storage/index/group_key/group_key_index.cpp

```diff
@@ -14,41 +15,60 @@ size_t GroupKeyIndex::estimate_memory_consumption(ChunkOffset row_count, ChunkOf
+  //    The ValueID for null in an AV is the highest available ValueID in the dictionary + 1.
+  //    `unique_values_count` returns the size of dictionary which does not store a ValueID for null.
+  //    Therefore we have `unique_values_count` + 1 (for null) ValueIDs for which we want to count the occurrences.
+  _index_offsets = std::vector<ChunkOffset>(
+      _indexed_segment->unique_values_count() + 1u /*for null*/ + 1u /*to mark the ending position */, 0u);
 
   // 2) Count the occurrences of value-ids: Iterate once over the attribute vector (i.e. value ids)
   //    and count the occurrences of each value id at their respective position in the dictionary,
   //    i.e. the position in the _index_offsets
-  resolve_compressed_vector_type(*_indexed_segments->attribute_vector(), [&](auto& attribute_vector) {
+  resolve_compressed_vector_type(*_indexed_segment->attribute_vector(), [&](auto& attribute_vector) {
     for (const auto value_id : attribute_vector) {
       _index_offsets[value_id + 1u]++;
     }
   });
 
-  // 3) Create offsets for the postings in _index_offsets
+  const auto null_value_id = _indexed_segment->unique_values_count();
+  const auto null_count = _index_offsets[null_value_id + 1u];
+  const auto non_null_count = _indexed_segment->size() - null_count;
+
+  // 3) Set the _index_postings and _index_null_postings
+  _index_postings = std::vector<ChunkOffset>(non_null_count);
+  _index_null_postings = std::vector<ChunkOffset>(null_count);
+
+  // 4) Create offsets for the postings in _index_offsets
   std::partial_sum(_index_offsets.begin(), _index_offsets.end(), _index_offsets.begin());
 
-  // 4) Create the postings
-  // 4a) Copy _index_offsets to use it as a write counter
-  auto index_offset_copy = std::vector<size_t>(_index_offsets);
+  // 5) Create the postings
+  // 5a) Copy _index_offsets to use it as a write counter
+  auto index_offset_copy = std::vector<ChunkOffset>(_index_offsets);
 
-  // 4b) Iterate once again over the attribute vector to obtain the write-offsets
-  resolve_compressed_vector_type(*_indexed_segments->attribute_vector(), [&](auto& attribute_vector) {
+  // 5b) Iterate once again over the attribute vector to obtain the write-offsets
+  resolve_compressed_vector_type(*_indexed_segment->attribute_vector(), [&](auto& attribute_vector) {
     auto value_id_it = attribute_vector.cbegin();
     auto position = 0u;
     for (; value_id_it != attribute_vector.cend(); ++value_id_it, ++position) {
       const auto& value_id = static_cast<ValueID>(*value_id_it);
-      _index_postings[index_offset_copy[value_id]] = position;
+
+      if (value_id != null_value_id) {
+        _index_postings[index_offset_copy[value_id]] = position;
+      } else {
+        const auto null_postings_offset = index_offset_copy[value_id] - non_null_count;
+        _index_null_postings[null_postings_offset] = position;
```

Comment:
I think this would be easier to read if you simply did held and incremented an iterator into `_index_null_postings`